### PR TITLE
PVS-Studio static analyzer warning fixes (part 2)

### PR DESCRIPTION
**Pull request by dimhotepus**
Part 2 of fixes by PVS-Studio static analyzer tool.

### CODE REVIEW

File: Externals/NVTT/src/nvimage/Filter.cpp [hunk 1]
```diff
@@ -329,7 +329,7 @@ Kernel2::Kernel2(const Kernel2 & k) : m_windowSize(k.m_windowSize)
 /// Dtor.
 Kernel2::~Kernel2()
 {
-	delete m_data;
+	delete [] m_data;
```
Comment (Xottab-DUTY) [inline Externals/NVTT/src/nvimage/Filter.cpp]:
```suggestion
	delete[] m_data;
```

File: Externals/NVTT/src/nvimage/Filter.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ Kernel1::Kernel1(const Filter & f, int iscale, int samples/*= 32*/)
 /// Dtor.
 Kernel1::~Kernel1()
 {
-	delete m_data;
+	delete [] m_data;
```
Comment (Xottab-DUTY) [inline Externals/NVTT/src/nvimage/Filter.cpp]:
```suggestion
	delete[] m_data;
```

File: src/editors/xrPostprocessEditor/EditorEngine.cs [hunk 3]
```diff
@@ -6,17 +6,27 @@
...
 
-    public struct ColorMappingParams
+    public struct ColorMappingParams : IEquatable<ColorMappingParams>
     {
         public float Influence;
         public string Texture;
+
+        public bool Equals(ColorMappingParams other)
+        {
+          return Influence == other.Influence && Texture == other.Texture;
```
Comment (Xottab-DUTY) [inline src/editors/xrPostprocessEditor/EditorEngine.cs]:
```suggestion
            return Influence == other.Influence && Texture == other.Texture;
```

File: src/Layers/xrRender/DetailManager.cpp [hunk 4]
```diff
@@ -213,16 +213,19 @@ void CDetailManager::Unload()
...
 
     objects.clear();
-    m_visibles[0].clear();
-    m_visibles[1].clear();
-    m_visibles[2].clear();
+
+    for (vis_list& l : m_visibles)
+    {
+        l.clear();
+    }
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/DetailManager.cpp]: Wastes too much space, introduces indentation to the right, too small – so, doesn't worth it.

File: src/utils/ETools/audio.cpp [hunk 5]
```diff
@@ -776,6 +792,7 @@ void setup_downmix(oe_enc_opt* opt)
     if (opt->channels != 2)
     {
         fprintf(stderr, "Internal error! Please report this bug.\n");
+        free(d);
```
Comment (Xottab-DUTY) [inline src/utils/ETools/audio.cpp]: It's better to just move the memory allocation down.

File: src/utils/ETools/audio.cpp [hunk 6]
```diff
@@ -250,6 +252,7 @@ int aiff_open(FILE* in, oe_enc_opt* opt, unsigned char* buf, int buflen)
     if (fread(buffer, 1, len, in) < len)
     {
         fprintf(stderr, _("Warning: Unexpected EOF in reading AIFF header\n"));
+        free(aiff);
```
Comment (Xottab-DUTY) [inline src/utils/ETools/audio.cpp]: We can just move the `aiff` allocation down to the place where it's actually used.

`aiff->bigendian =` can be replaced with `int bigendian =`

File: src/utils/ETools/audio.cpp [hunk 7]
```diff
@@ -366,11 +375,15 @@ int wav_open(FILE* in, oe_enc_opt* opt, unsigned char* oldbuf, int buflen)
      */
 
     if (!find_wav_chunk(in, "fmt ", &len))
+    {
+        free(wav);
```
Comment (Xottab-DUTY) [inline src/utils/ETools/audio.cpp]: We can just move the `wav` allocation down to the place where it's actually used.

File: src/Include/editor/ide.hpp [hunk 8]
```diff
@@ -22,6 +22,7 @@ class property_holder_holder;
 class ide_base
 {
 public:
+    virtual ~ide_base() {}
```
Comment (Xottab-DUTY) [inline src/Include/editor/ide.hpp]:
```suggestion
    virtual ~ide_base() = default;
```

File: src/Include/editor/property_holder_base.hpp [hunk 9]
```diff
@@ -126,6 +126,7 @@ class property_holder_base
     }; // enum value_editor_enum
 
 public:
+    virtual ~property_holder_base() {}
```
Comment (Xottab-DUTY) [inline src/Include/editor/property_holder_base.hpp]:
```suggestion
    virtual ~property_holder_base() = default;
```

File: src/editors/xrSdkControls/Controls/TreeViewSearchPanel/TreeViewSearchPanel.cs [hunk 10]
```diff
@@ -72,7 +72,7 @@ private void searchBox_KeyDown(object sender, KeyEventArgs e)
                     MessageBox.Show("There is no matches!");
                 else if (findedNodes.Count == 1)
                     resultLabel.Text = "Result: 1 / 1";
-                else if (findedNodes.Count > 1)
```
Comment (Xottab-DUTY) [inline src/editors/xrSdkControls/Controls/TreeViewSearchPanel/TreeViewSearchPanel.cs:75]: And now with have two `else { ... } else { ... }` branches

File: src/Layers/xrRender/DetailManager.cpp (hunk 4 above)
Comment (dimhotepus) [reply src/Layers/xrRender/DetailManager.cpp]: Ok, restored.

File: src/utils/ETools/audio.cpp (hunk 6 above)
Comment (dimhotepus) [reply src/utils/ETools/audio.cpp]: Unfortunately, `aiff` is used in nested block, too. 

May be use `std::make_unique` + `release` where needed?

File: src/editors/xrSdkControls/Controls/TreeViewSearchPanel/TreeViewSearchPanel.cs (hunk 10 above)
Comment (dimhotepus) [reply src/editors/xrSdkControls/Controls/TreeViewSearchPanel/TreeViewSearchPanel.cs:75]: Sorry, output window has hidden last else from me. Just removed dead code.

File: src/utils/ETools/audio.cpp (hunk 6 above)
Comment (dimhotepus) [reply src/utils/ETools/audio.cpp]: Nevermind, fixed by `int bigendian =`

File: src/utils/ETools/audio.cpp [hunk 11]
```diff
@@ -366,7 +366,9 @@ int wav_open(FILE* in, oe_enc_opt* opt, unsigned char* oldbuf, int buflen)
      */
 
     if (!find_wav_chunk(in, "fmt ", &len))
+    {
         return 0; /* EOF */
+    }
```
Comment (Xottab-DUTY) [inline src/utils/ETools/audio.cpp]: Unneeded formatting

File: src/Layers/xrRender/DetailManager.cpp [hunk 12]
```diff
@@ -213,16 +213,17 @@ void CDetailManager::Unload()
...
-        xr_delete(*it);
+        o->Unload();
+        xr_delete(o);
     }
 
     objects.clear();
     m_visibles[0].clear();
     m_visibles[1].clear();
     m_visibles[2].clear();
+
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/DetailManager.cpp]: Unneeded formatting

File: src/utils/ETools/audio.cpp [hunk 13]
```diff
@@ -258,7 +257,7 @@ int aiff_open(FILE* in, oe_enc_opt* opt, unsigned char* buf, int buflen)
     format.samplesize = READ_U16_BE(buffer + 6);
     format.rate = (int)read_IEEE80(buffer + 8);
 
-    aiff->bigendian = 1;
+    int bigendian = 1;
```
Comment (Xottab-DUTY) [inline src/utils/ETools/audio.cpp]: Was wrong a bit. Let's make it 100% correspond to `aifffile::bigendian` type
```suggestion
    short bigendian = 1;
```

File: src/Layers/xrRender/DetailManager.cpp [hunk 14]
```diff
@@ -213,16 +213,17 @@ void CDetailManager::Unload()
 {
     hw_Unload();
 
-    for (DetailIt it = objects.begin(); it != objects.end(); it++)
+    for (CDetail* o : objects)
```
Comment (Xottab-DUTY) [inline src/Layers/xrRender/DetailManager.cpp]: It's better to name it `detailObject`.

**Comment by dimhotepus** [comment]
Not sure about Codacity's `Do not check floating point equality with exact values, use a range instead.` issues.

I can add correct implementation of floating point comparison via `Math.Abs(expected - actual) <|<= Epsilon`, but two questions arise:
* Which `Epsilon` should i choose?
* Do we need this at all for now?

**Comment by Xottab-DUTY** [comment]
[OpenXRay#666@aba138b](https://github.com/OpenXRay/xray-16/pull/666/commits/aba138b183420cb5165f7f87bf3e24248a29408c)
It would be unhandy to rebase this commit 😬

**Comment by dimhotepus** [comment]
> [OpenXRay#666@aba138b](https://github.com/OpenXRay/xray-16/pull/666/commits/aba138b183420cb5165f7f87bf3e24248a29408c)
> It would be unhandy to rebase this commit 😬

May be do not rebase? Or you prefer `every commit is a release one` idiom?

**Comment by Xottab-DUTY** [comment]
> May be do not rebase? Or you prefer every commit is a release one idiom?

Not exactly, but almost like that. I'm just trying to keep history clean.
The rule sounds something like this: `once a WIP branch is ready to merge, it should be cleaned up from commits that fix previous commits. (fixing commits should be fixup'ed into the commits they are fixing)`

> Not sure about Codacity's `Do not check floating point equality with exact values, use a range instead.` issues.
> 
> I can add correct implementation of floating point comparison via `Math.Abs(expected - actual) <|<= Epsilon`, but two questions arise:
> 
> * Which `Epsilon` should i choose?
> * Do we need this at all for now?

Not sure about this too... Probably, we can ignore the Codacy's warning... But, when IEquatable was unimplemented and reflection was used, how it was implemented? Was it checking exact values or using a proper comparison with epsilon?

**Comment by dimhotepus** [comment]
> > May be do not rebase? Or you prefer every commit is a release one idiom?
> 
> Not exactly, but almost like that. I'm just trying to keep history clean.
> The rule sounds something like this: `once a WIP branch is ready to merge, it should be cleaned up from commits that fix previous commits. (fixing commits should be fixup'ed into the commits they are fixing)`

Ok, i'll split and rebase.

> > Not sure about Codacity's `Do not check floating point equality with exact values, use a range instead.` issues.
> > I can add correct implementation of floating point comparison via `Math.Abs(expected - actual) <|<= Epsilon`, but two questions arise:
> > 
> > * Which `Epsilon` should i choose?
> > * Do we need this at all for now?
> 
> Not sure about this too... Probably, we can ignore the Codacy's warning... But, when IEquatable was unimplemented and reflection was used, how it was implemented? Was it checking exact values or using a proper comparison with epsilon?

Hm. If i remember correctly, C# spec says value types default equality operator uses bitwise comparison, so behavior is the same as before except costly reflection calls are replaced by our custom `Equals` implementation.

**Comment by Xottab-DUTY** [comment]
> Hm. If i remember correctly, C# spec says value types default equality operator uses bitwise comparison, so behavior is the same as before except costly reflection calls are replaced by our custom Equals implementation.

Then we can ignore the warning.

**Comment by dimhotepus** [comment]
> > Hm. If i remember correctly, C# spec says value types default equality operator uses bitwise comparison, so behavior is the same as before except costly reflection calls are replaced by our custom Equals implementation.
> 
> Then we can ignore the warning.

Thank you for review. I've rebased and squashed all the things.

File: src/editors/xrWeatherEditor/property_converter_color.cpp [hunk 15]
```diff
@@ -83,7 +83,7 @@ bool property_converter_color::CanConvertFrom(ITypeDescriptorContext ^ context,
 Object ^ property_converter_color::ConvertFrom(ITypeDescriptorContext ^ context, CultureInfo ^ culture, Object ^ value)
 {
     String ^ string = dynamic_cast<String ^>(value);
-    if (!value)
```
Comment (Xottab-DUTY) [inline src/editors/xrWeatherEditor/property_converter_color.cpp:86]: I just checked the code and I see that this replacement is wrong and the code was written like that intentionally.

Comment (Xottab-DUTY) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: But it makes sense to swap the code like that:
```
	if (!value)
		return (inherited::ConvertFrom(context, culture, value));

	String ^ string = dynamic_cast<String ^>(value);
	try
	{
		...
```

Comment (dimhotepus) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: Good to catch it before the merge.

I'll move cast below value check than.

Comment (dimhotepus) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: > But it makes sense to swap the code like that:
> 
> ```
> 	if (!value)
> 		return (inherited::ConvertFrom(context, culture, value));
> 
> 	String ^ string = dynamic_cast<String ^>(value);
> 	try
> 	{
> 		...
> ```

Done, thank you. Hope space doesn't matter here.

Comment (Xottab-DUTY) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: Well, it doesn't look good, sorry
oops

Comment (dimhotepus) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: It'll be better to use ClangFormat or similar tool to automate boring formatting task.

I'll fix formatting tomorrow, thanks!

Comment (dimhotepus) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: Done!

Comment (Xottab-DUTY) [reply src/editors/xrWeatherEditor/property_converter_color.cpp:86]: Thank you!

> It'll be better to use ClangFormat or similar tool to automate boring formatting task.

We have `.clang-format` file in the `src` dir. Doesn't it work?

**Review by Xottab-DUTY — APPROVED** [review]

**Comment by Xottab-DUTY** [comment]
Thank you!

**Pull request merged**